include/rocksdb/env_encrypt2.h

```diff
@@ -0,0 +1,321 @@
+    } else {
+      valid = false;
+    }
+  }
+
+  AesCtrKey(const std::string& key_str);
+
+  // see Writing Solid Code, 2nd edition
+  //   Chapter 9, page 321, Managing Secrets in Memory ... bullet 4 "Scrub the
+  //   memory"
+  // Not saying this is essential or effective in initial implementation since
+  // current
+  //  usage model loads all keys at start and only deletes them at shutdown. But
+  //  does establish presidence.
+  // goal is to explicitly remove key from memory once no longer needed
+  ~AesCtrKey() {
+    memset(key, 0, EVP_MAX_KEY_LENGTH);
+    valid = false;
+  }
+
+  bool operator==(const AesCtrKey& rhs) const {
+    return (0 == memcmp(key, rhs.key, EVP_MAX_KEY_LENGTH)) &&
+           (valid == rhs.valid);
+  }
+
+  bool IsValid() const { return valid; }
+};
+
+class CTREncryptionProviderV2 : public EncryptionProvider {
+ public:
+  CTREncryptionProviderV2() = delete;
+
+  CTREncryptionProviderV2(const CTREncryptionProvider&&) = delete;
+
+  CTREncryptionProviderV2(const Sha1Description& key_desc_in,
+                          const AesCtrKey& key_in)
+      : valid_(false), key_desc_(key_desc_in), key_(key_in) {
+    valid_ = key_desc_.IsValid() && key_.IsValid();
+  }
+
+  CTREncryptionProviderV2(const std::string& key_desc_str,
+                          const uint8_t unformatted_key[], int bytes)
+      : valid_(false), key_desc_(key_desc_str), key_(unformatted_key, bytes) {
+    valid_ = key_desc_.IsValid() && key_.IsValid();
+  }
+
+  size_t GetPrefixLength() const override;
+
+  Status CreateNewPrefix(const std::string& /*fname*/, char* prefix,
+                         size_t prefixLength) const override;
+
+  Status CreateCipherStream(
+      const std::string& /*fname*/, const EnvOptions& /*options*/,
+      Slice& /*prefix*/,
+      std::unique_ptr<BlockAccessCipherStream>* /*result*/) override {
+    return Status::NotSupported("Wrong EncryptionProvider assumed");
+  }
+
+  virtual BlockAccessCipherStream* CreateCipherStream2(
+      uint8_t code_version, const uint8_t nonce[]) const;
+
+  bool Valid() const { return valid_; };
+  const Sha1Description& key_desc() const { return key_desc_; };
+  const AesCtrKey& key() const { return key_; };
+
+ protected:
+  bool valid_;
+  Sha1Description key_desc_;
+  AesCtrKey key_;
+};
+
+class EncryptedWritableFileV2 : public EncryptedWritableFile {
+ public:
+  // Default ctor. Prefix is assumed to be written already.
+  EncryptedWritableFileV2(std::unique_ptr<WritableFile>&& f,
+                          std::unique_ptr<BlockAccessCipherStream>&& s,
+                          size_t prefix_length)
+      : EncryptedWritableFile(std::move(f), std::move(s), prefix_length) {}
+
+  Status Append(const Slice& data) override;
+
+  Status PositionedAppend(const Slice& data, uint64_t offset) override;
+
+  // Indicates the upper layers if the current WritableFile implementation
+  // uses direct IO.
+  bool use_direct_io() const override { return false; };
```

Comment:
Is there a reason direct IO does not work for these files?

Reply:
Alignment.  These two types of files have a different memory alignment and I lacked the time and experience to verify the impact.